adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1534,43 +1534,53 @@ protected final void onCommandReceived(
+                                            targetInfo, subscription.getTenant(), endpoint.clientIdentifier(),
+                                            subscription.getQos(), publishTopic, thr);
+                                    TracingHelper.logError(commandContext.getTracingSpan(), "failed to publish command", thr);
+                                    reportPublishedCommand(
+                                            tenantObject,
+                                            subscription,
+                                            commandContext,
+                                            ProcessingOutcome.from(thr));
+                                    commandContext.release();
+                                });
+                    }).onFailure(t -> {
+                        log.debug("error mapping command [tenant-id: {}, MQTT client-id: {}, QoS: {}]",
+                                subscription.getTenant(), endpoint.clientIdentifier(), subscription.getQos(), t);
+                        TracingHelper.logError(commandContext.getTracingSpan(), "failed to map command", t);
+                        reportPublishedCommand(
                                 tenantObject,
                                 subscription,
                                 commandContext,
-                                ProcessingOutcome.from(sentHandler.cause()));
-                            commandContext.release();
-                        }
+                                ProcessingOutcome.from(t));
+                        commandContext.release();
                     });
-                }
-            ).onFailure(t -> {
-                log.debug("error mapping command [tenant-id: {}, MQTT client-id: {}, QoS: {}]",
-                    subscription.getTenant(), endpoint.clientIdentifier(),
-                    subscription.getQos(), t);
-                TracingHelper.logError(commandContext.getTracingSpan(), "failed to map command", t);
-                reportPublishedCommand(
-                    tenantObject,
-                    subscription,
-                    commandContext,
-                    ProcessingOutcome.from(t));
-                commandContext.release();
-            });
+        }
+
+        private Future<Integer> publish(final String topic, final Buffer payload, final MqttQoS qosLevel) {
+            final Promise<Integer> publishSentPromise = Promise.promise();
+            try {
+                endpoint.publish(topic, payload, qosLevel, false, false, publishSentPromise);
+            } catch (final Exception e) {
+                publishSentPromise.fail(!endpoint.isConnected()
+                        ? new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "connection to device already closed")
+                        : new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR, e));
```

Comment:
Isn't it better to check if device is connected or not even before publishing instead of catching the exception and then categorizing it?

Reply:
The case that the endpoint may not be connected here is already rather exceptional (means the disconnect happened right *after* the command was received by the adapter but *before* the extra checks before forwarding the command to the device were completed). Therefore I prefer the solution above to adding an extra `isConnected` check (which is synchronized in `MqttEndpointImpl`) for each `publish` invocation here.